[Enhancement] Support zstd level

srlch opened the pull request:
In this pr, we introduce configable compression level for ZSTD compression type.
Valid level is 1-22, default level is 3 (which is same as before)

How to use:
CREATE TABLE t ( k BIGINT NOT NULL, v string COMMENT) ENGINE=OLAP 
DUPLICATE KEY(k)
DISTRIBUTED BY HASH(k) BUCKETS 1
PROPERTIES ( \"compression\" = \"zstd(1)\" );

We can define the compression level for zstd using zstd(level) in CREATE TABLE STATEMENT

We load 10000000row from Clickbench dataset, using different level of zstd compression, result as following:
| level | show data |
|----------|----------|
| 1  | 103.477 MB   |
| 3 (defualt)   | 100.557 MB  |
| 5   | 98.896 MB |
| 10   | 96.549 MB   |
| 15   | 95.890 MB   |
| 20   | 91.921 MB  |
| 22   | 91.921 MB   |

Fixes https://github.com/StarRocks/starrocks/issues/46839

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

kevincai commented on be/src/storage/tablet_meta.cpp:
```diff
@@ -121,6 +121,10 @@ TabletMeta::TabletMeta(int64_t table_id, int64_t partition_id, int64_t tablet_id
     auto st = convert_t_schema_to_pb_schema(tablet_schema, next_unique_id, col_ordinal_to_unique_id, schema,
                                             compression_type);
+    // compression level is only used for zstd for now.
+    if (compression_type == TCompressionType::ZSTD && compression_level != -1) {
```
what if the compress level is set anyway for other compression type? doesn't hurt anything, right?

kevincai commented on be/src/util/compression/block_compression.cpp:
```diff
@@ -753,6 +767,21 @@ class ZstdBlockCompression final : public BlockCompressionCodec {
+        // pool by reseting back to level = ZSTD_CLEVEL_DEFAULT(3).
+        // What we should do here is simply set the level as we wanted.
+        if (_level != -1) {
+            DCHECK(_level >= 1 && _level <= 22);
```
DHCECK is not good enough. In release mode it is nothing. shall fail and error out if the level is beyond [1, 22],

srlch replied:
Fixed

kevincai commented on be/src/util/compression/block_compression.cpp:
```diff
@@ -718,13 +718,27 @@ class ZlibBlockCompression : public BlockCompressionCodec {
     }
 
+    static const ZstdBlockCompression* instance(int level) {
+        DCHECK(level >= 1 && level <= 22);
```
remove this DCHECK and return nullptr if level is out of range.

Never take DCHECK as a the parameter error handling.

srlch replied:
Fixed

srlch committed:
[Enhancement] Support zstd level

In this pr, we introduce configable compression level for ZSTD compression type.
Valid level is 1-22, default level is 3 (which is same as before)

How to use:
CREATE TABLE t ( k BIGINT NOT NULL, v string COMMENT) ENGINE=OLAP
DUPLICATE KEY(k)
DISTRIBUTED BY HASH(k) BUCKETS 1
PROPERTIES ( "compression" = "zstd(0)" );

We can define the compression level for zstd using zstd(level) in CREATE TABLE STATEMENT

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Fxi sql test

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented on fe/fe-core/src/main/java/com/starrocks/catalog/TableProperty.java:
```diff
@@ -605,6 +608,25 @@ public TableProperty buildCompressionType() {
                 return this;
             }
         }
+        String noSpacesCompression = compression.replace(" ", "");
```
Fixed

srlch committed:
fix format

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix format

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

wyb commented on fe/fe-core/src/main/java/com/starrocks/catalog/TableProperty.java:
```diff
@@ -244,6 +244,9 @@ public static String valueList() {
     private TCompressionType compressionType = TCompressionType.LZ4_FRAME;
 
+    // the default compression level of this table, only used for zstd for now.
+    private int compressionLevel = -1;
```
`compressionLevel` is serialized in `properties`?

wyb commented on be/src/util/compression/block_compression.cpp:
```diff
@@ -718,13 +718,29 @@ class ZlibBlockCompression : public BlockCompressionCodec {
     }
 
+    static const ZstdBlockCompression* instance(int level) {
+        if (level <= 0 || level >= 23) {
```
```suggestion
        if (level < 1 || level > 22) {
```

wyb replied:
same as line 779

srlch replied:
Fixed

srlch committed:
fix comment

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Fix NPE

Signed-off-by: srlch <linzichao@starrocks.com>

kevincai commented:
@srlch is this needed for branch-3.3?

srlch commented:
@kevincai The release plan of this feature has not yet been determined. It may be backported to 3.3 if it is confirmed.

srlch commented:
https://github.com/Mergifyio backport branch-3.3
